### DISCUSSION TITLE
Resource-Based API

### DISCUSSION THREAD

**Discussion by livio-a** [zitadel/zitadel#5875] [Ideas]
We're changing the structure of our API. Previously, our APIs were organized by UseCase/Context, such as Auth API for authenticated users and Management API for organization managers. However, this led to confusion about which API to use, particularly for requests that could be useful across multiple APIs but with different filters. For instance, SearchUsers on an Instance Level or on an Organization Level.

To address this issue, we're implementing a new resource-based API. With this approach, we have one API with services for different resources (e.g., User, Session, Settings). Depending on the requesting user's permissions, they're allowed to request more or less. We're starting with Users and Session resources, so you can build your own login UI, giving you greater customization flexibility.

Our approach is an MVP style. Initially, we'll allow for creating and logging in users with usernames and passwords. Later on, we'll add more features, like multifactor and identity providers, until we cover the existing possibilities. Track our progress in this [issue](https://github.com/zitadel/zitadel/issues/5015).

And now we need your feedback. First endpoints for user creation and session management are implemented and we want to know what you think about it.

[User Service](https://zitadel.com/docs/category/apis/resources/user_service/user-service)
[Session Service](https://zitadel.com/docs/category/apis/resources/session_service/session-service)
[Settings Service](https://zitadel.com/docs/category/apis/resources/settings_service/settings-service)
The user service and settings service are mostly self describing. The user service is for creating new users, changing and verifying email addresses. In the future there will also be functionalities like password reset, change user, add phone, etc be added.
The settings service is used to retrieve all configurations you have made in ZITADEL like login, password, branding settings. For now we have only implemented the query part, to allow you to implement your login ui completely with the new API.

### Session Service
The session service allows you to build your own login flow. Here are the thoughts we had, and this is also the part where we look forward to your feedback the most.
[Create](https://zitadel.com/docs/apis/resources/session_service/session-service-create-session) and [update](https://zitadel.com/docs/apis/resources/session_service/session-service-set-session) session do almost look the same. This gives you the possibility to either create an empty session on opening up a new login ui and enrich the session while the user is going through the login flow, or you can directly create the session with for example the username and password that have to be checked.
These two endpoints will always return a token where you find all the relevant information about the session and you are able to decide whether the user is done or not with the login. 

When creating or updating a session you have two different attributes, the checks and the metadata. The checks are handled by ZITADEL and allow you to authenticate your user. E.g username check, password check, etc. The metadata is a key value list, which allows you to add more data to the session so you can eventually do your own checks on specific data.

```
{
  "checks": {
    "user": {
      "loginName": "mini@mouse.com"
    },
    "password": {
      "password": "V3ryS3cure!"
    }
  },
  "metadata": {
    “custom-key”: c29tZS12YWx1ZQ==
  }
}
```

So let us know what you think about the approach? How do you like the API structure? What do you think about the requests and responses? Any feedback is welcome.

Note: We recently released the new API in an Alpha state, which can lead to breaking changes. We will stay in this state, till we have built all the needed resources to cover our current login flow.

Read the following discussion to find out how we will handle the UI: https://github.com/zitadel/zitadel/discussions/5922

**Comment by Alexei-Barnes** [comment]
Great stuff!

The request/response body structures look solid as of first read.

The sessions service offers username/password login, but what are the plans for how a developer would build a login flow here with MFA/Passwordless/SSO? Perhaps it would look like an expansion to the `.checks` part of the create session request body? For example, if I have `forceMFA` set to `true` then presumably, there needs to be some way to perform that check, either in create or update.

**Reply by livio-a** [reply]
Hey @Alexei-Barnes 

We're currently adding the possibility to start an IDP flow (for registration and soon for login): https://github.com/zitadel/zitadel/pull/5879
The login part will as you suggested be solved with a `check`.
The same applies to passkeys (passwordless) or other 2FA/MFA authentications, like (T)OTP. Some might also require two steps, where you will need to do a call first to get the required information for the `check` itself.
So for example for the passkey authentication you will call a `CreatePasskeyAuthentication` endpoint where you get the necessary information for the WebAuthN verification for the browser as well as a token, which will also be required in the `PasskeyCheck` to update the Session.
TOTP on the other hand can be solved with a `check` directly and won't need an additional call first.

Is that about what you imagined too?

**Reply by Alexei-Barnes** [reply]
Thanks, yea I think that's a reasonable API. It should certainly be possible to build a login UI ontop of that.

**Comment by Sipioteo** [comment]
I'm trying to use the API but the doc is a bit confusing.
It says that it can be used to create your own "login ui" but i can't find how to convert a sessionToken to accessToken+refreshToken+idToken.

Can you provide some more explanations?

**Reply by livio-a** [reply]
Hey @Sipioteo 

We still need to solve / implement that part: https://github.com/zitadel/zitadel/issues/5468
The basic idea is, that you can either use the session token directly as opaque access_token or most likely to have an endpoint to exchange the session token with the tokens.
We will provide SDKs to help build your own Login UI (starting with one for typescript: https://github.com/zitadel/typescript), which will provide the functionality to do so.

We will update the linked issue and docs as soon as we have a working draft.

**Reply by Sipioteo** [reply]
Well, i tried to use the session token as an access token but zitadel replyes "Wrong token" when i do request like /oidc/v1/userinfo or /oauth/v2/introspect. I literally not have a way to check if the token is valid or not.

**Comment by reactima** [comment]
Session service and custome login flow looks like very powerful move. 

After exploring @zitadel/server code with v2 functionality in 

https://github.com/zitadel/typescript
https://github.com/zitadel/typescript/tree/main/packages/zitadel-server/src/v2

I stared wondering about https://github.com/zitadel/zitadel-go 

Will zitadel-go be synced properly with proto files and has the same functionality with @zitadel/server? 

I'm asking cause zitadel-go is already missing v2 functionality. I'm mainly using golang on backend would like to avoid nextjs backend ecosystem. 

Most of my other thoughts are spinning about possible overcommitment to nextjs ecosystems. 
It would be nice to see sample apps with golang, python based session management on backend and UI components from https://github.com/zitadel/typescript/tree/main/packages/zitadel-react

**Reply by fforootd** [reply]
Good question.

The reason why we started with a nextjs project is actually to better cover the js/ts ecosystem and to give frontend devs some more flexibility. In this process we are implementing the current Login UI of ZITADEL into a nextjs app (we like eat your own dog food).

We want our new APIs to be used in a wider audience so we will extend the SDKs to cover the v2 APIs for sure. What will be out of scope is to build a predefined and reusable login UI in go 😁 

@hifabienne @livio-a do you see are possibility that we update the zitadel-go sdk to a version that supports v2 as well?
To me it looks like we would need to extend this script here https://github.com/zitadel/zitadel-go/blob/main/build/zitadel/generate-grpc-client.sh#L17 or replace it with buf 😁

**Reply by livio-a** [reply]
Yes we need to extend that script / replace it with buf and probably something in the Dockerfile. I'll try to have a look at it this week.

**Comment by hifabienne** [comment]
You can find more information about how to build your own login UI with our APIs in our published documentation, as always feedback is highly appreciated: https://zitadel.com/docs/guides/integrate/login-ui

**Reply by geekyme-fsmk** [reply]
Thanks! I had a discovery call with Max and one of the things I'm trying to do is to have my own Login UI that rides on top of zitadel APIs. 

1. I need to have OTP codes sent out to my users on login, is there a way to do that by using your system to generate the code, and then sending the code out myself with my own twilio integration? 

2. I have other forms of login such as biometric / FaceID login. How would that work?

**Reply by hifabienne** [reply]
Hei @geekyme-fsmk 
1. This functionality is currently not implemented. I have created an issue here: https://github.com/zitadel/zitadel/issues/6100 Please upvote if you like to see this implemented.
2. Passkey is already implemented in the API, but currently the docs in the guide are missing. Generally it works similar than the identity providers. You have to start a check first, and than add this to the session of your user. 
- Start Passkey Registration: https://zitadel.com/docs/apis/resources/user_service/user-service-register-passkey
- Verify Passkey Registration: https://zitadel.com/docs/apis/resources/user_service/user-service-verify-passkey-registration
- Create or update a session with the passkey https://zitadel.com/docs/apis/resources/session_service/session-service-create-session 
https://zitadel.com/docs/apis/resources/session_service/session-service-set-session

We will add the passkey registration docs asap
